src/controllers/projectController.ts

```diff
@@ -94,4 +108,18 @@ export class ProjectController {
         }
         return javaVersion;
     }
+
+    private async refreshExcludeFilters() {
+        const patterns: string[] = [];
+        const excludeSetting = workspace.getConfiguration("files", null).get<{ [pattern: string]: boolean }>("exclude");
+        for (const pattern of  Object.keys(excludeSetting)) {
```

Comment:
Did you try to exclude ".classpath, .project, .settings"? What happens?

Reply:
Before excluding:
<img src=https://user-images.githubusercontent.com/20227484/70215602-c6f9de80-1778-11ea-9ee6-b83b43faf010.png width=300 height=500>

After excluding in workspace:
![image](https://user-images.githubusercontent.com/20227484/70215681-fc9ec780-1778-11ea-978c-9f19435772a0.png)

The packagve view automatically refreshed.

Reply:
It reminds me of another issue:

The project won't delete the filter when corresponding exclude settings got removed.

Reply:
> It reminds me of another issue:
> 
> The project won't delete the filter when corresponding exclude settings got removed.

This issue has been solved in new commit.

Reply:
From the implementation, it actually adds a filter to project description so as to be excluded JDT resource model. I'm worried that if .project and .classpath is excluded from the resource model, how did JDT understand your project?

Reply:
I debugged this situation for a while, and everything works quite well.

Try to clarify what happened:
1. `.project`, `.class` are normally used by `JavaProject` class.
2. Take [`JavaProject.getRawClasspath()`](https://github.com/eclipse/eclipse.jdt.core/blob/f6792492f110837cda3b566a10e36beea4e54b03/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java#L2354) as example, it fetchs project info from [`getPerProjectInfo()`](https://github.com/eclipse/eclipse.jdt.core/blob/f6792492f110837cda3b566a10e36beea4e54b03/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java#L2325) method (and many other methods also uses it)
3.  `JavaProject.getPerProjectInfo()` indirectly calls [`JavaModelManager.getPerProjectInfo(IProject project, boolean create = false)`](https://github.com/eclipse/eclipse.jdt.core/blob/f6792492f110837cda3b566a10e36beea4e54b03/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java#L2537)
4. `getPerProjectInfo` checks the existence project in its member [`Map<IProject, PerProjectInfo> this.perProjectInfos`](https://github.com/eclipse/eclipse.jdt.core/blob/f6792492f110837cda3b566a10e36beea4e54b03/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java#L2537).
5. `PerProjectInfo` class reads `.classpath` file via [`JavaProject.readFileEntriesWithException`](https://github.com/eclipse/eclipse.jdt.core/blob/f6792492f110837cda3b566a10e36beea4e54b03/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java#L2915)

---

Here:
`JavaProject.readFileEntriesWithException` uses `IProject.getFile` to get classpath resource:

```java
IFile rscFile = this.project.getFile(JavaProject.CLASSPATH_FILENAME);
```

Here, `IProject.getFile` **will** return the filtered file:
```java
> javaProject.getProject().getFile(javaProject.CLASSPATH_FILE_NAME)
File@409 "L/microrts_c32e57f8/.classpath"
> javaProject.getProject().getFile(".DS_Store")
File@442 "L/microrts_c32e57f8/.DS_Store"
```
So it could properly work.

Reply:
Additional output:
```java
> javaProject.getProject().getWorkspace().validateFiltered(javaProject.getProject().getFile("_/src/ai/.DS_Store"))
ResourceStatus@615 "[type: INFO], [path: null], [message: The resource will be filtered out by its parent resource filters], [plugin: org.eclipse.core.resources], [exception: null]\n"
> javaProject.getProject().getWorkspace().validateFiltered(javaProject.getProject().getFile(".classpath"))
ResourceStatus@621 "[type: INFO], [path: null], [message: The resource will be filtered out by its parent resource filters], [plugin: org.eclipse.core.resources], [exception: null]\n"
```